includes/class.llms.generator.php

```diff
@@ -21,892 +21,211 @@
-			if ( isset( $raw[ $key ] ) ) {
-				$question->set( $key, $raw[ $key ] );
-			}
-		}
-
-		do_action( 'llms_generator_new_question', $question, $raw, $manager, $this );
-
-		return $question->get( 'id' );
-
-	}
-
-	/**
-	 * Creates a new section
-	 *
-	 * Creates all lessons within the section data.
-	 *
-	 * @since 3.3.0
-	 * @since 3.30.2 Added hooks.
-	 *
-	 * @param array $raw                Raw section data.
-	 * @param int   $order              Order within the course (starts at 1).
-	 * @param int   $course_id          WP Post ID of the parent course.
-	 * @param int   $fallback_author_id Optional author ID to use as a fallback if no raw author data supplied for the lesson.
-	 * @return int The WP_Post ID of the generated section.
-	 */
-	private function create_section( $raw, $order, $course_id, $fallback_author_id = null ) {
-
-		$raw = apply_filters( 'llms_generator_before_new_section', $raw, $order, $course_id, $fallback_author_id, $this );
-
-		$author_id = $this->get_author_id_from_raw( $raw, $fallback_author_id );
-
-		// Insert the course.
-		$section = new LLMS_Section(
-			'new',
-			array(
-				'post_author'   => $author_id,
-				'post_date'     => isset( $raw['date'] ) ? $this->format_date( $raw['date'] ) : null,
-				'post_modified' => isset( $raw['modified'] ) ? $this->format_date( $raw['modified'] ) : null,
-				'post_status'   => isset( $raw['status'] ) ? $raw['status'] : $this->get_default_post_status(),
-				'post_title'    => $raw['title'],
-			)
-		);
-
-		if ( ! $section->get( 'id' ) ) {
-			return $this->error->add( 'section_creation', __( 'Error creating section', 'lifterlms' ) );
 		}
 
-		$this->increment( 'sections' );
-
-		$section->set( 'parent_course', $course_id );
-		$section->set( 'order', $order );
-
-		if ( isset( $raw['lessons'] ) ) {
-			foreach ( $raw['lessons'] as $lesson_order => $lesson ) {
-				$this->create_lesson( $lesson, $lesson_order + 1, $section->get( 'id' ), $course_id, $author_id );
-			}
-		}
+		global $wpdb;
 
-		do_action( 'llms_generator_new_section', $section, $raw, $this );
+		$wpdb->hide_errors();
 
-		return $section->get( 'id' );
+		$wpdb->query( 'START TRANSACTION' ); // db call ok; no-cache ok.
 
-	}
+		/**
+		 * Action run immediately prior to a LifterLMS Generator running.
+		 *
+		 * @since 3.30.2
+		 *
+		 * @param LLMS_Generator $generator The generator instance.
+		 */
+		do_action( 'llms_generator_before_generate', $this );
 
-	/**
-	 * Ensure raw dates are correctly formatted to create a post date
-	 *
-	 * Falls back to current date if no date is supplied.
-	 *
-	 * @since 3.3.0
-	 * @since 3.30.2 Made publicly accessible.
-	 *
-	 * @param string $raw_date Raw date from raw object.
-	 * @return string
-	 */
-	public function format_date( $raw_date = null ) {
-
-		if ( ! $raw_date ) {
-			return current_time( 'mysql' );
-		} else {
-			return date( 'Y-m-d H:i:s', strtotime( $raw_date ) );
+		try {
+			call_user_func( $this->generator, $this->raw );
```

Comment:
```suggestion
			( $this->generator )( $this->raw );
```
why don't we use just this? ^

Reply:
Why?

Reply:
I should give you links about benchmarks of that function, I know, although we already know that function calls have a cost in PHP, so if we could simply avoid it, why not?